alchemist-engine/src/main/java/it/unibo/alchemist/core/implementations/Engine.java

```diff
@@ -345,12 +346,12 @@ public void play() {
     }
 
     @Override
-    public void reactionAdded(final Reaction<T> reactionToAdd) {
+    public void reactionAdded(final GlobalReaction<T> reactionToAdd) {
         reactionChanged(new ReactionAddition(reactionToAdd));
```

Comment:
We do need to fix this

Reply:
I reviewed this. `run` is protected by a lock on `environment`, and `myThread` is set once. calls to `reactionAdded` (hence, to `checkCaller`) ran before the assignment would compare with `null` and throw `IllegalMonitorStateException`. If the calling thread is not the simulation thread, we DO want the exception, so even if there is a race condition, it is fine (it would enter the condition anyway). For a race that produces unwanted results to occur (namely: the calling thread is the simulation thread, but an exception is raised), we would need the simulation thread to be both setting myThread and reading it, which is impossible.

I believe this is a false positive, though -- @kelvin-olaiya ? @cric96 @metaphori guys what do you think? Can you give us an opinion?

Reply:
TL;DR I check the code and I agree with @danysk: I think that this is a false positive.

The only potential harmful situation that popup in my mind is when you have a concurrent thread that tries to add remove/add a reaction before "myThread" is assigned (probably simulation thread is a better choice for the name??). Anyway, in this case, as @danysk says, the variable is still `null`, so the check caller will be thrown the exception, that is what I excepted will happen...

Reply:
I also agree with both @DanySK and @cric96. The `checkCaller` call prevents the race condition. If any changes in the simulation must be performed by `myThread`,  any concurrent thread trying to remove/add reaction should and will throw the Exception. 
At this point, I also agree with @cric96 on the thread naming. I will rename it to `simulationThread` as soon I'll get back to work.

Reply:
Very good. Marking as solved.